packages/network-of-terms-query/src/lookup/lookup.ts

```diff
@@ -69,15 +72,36 @@ export class LookupService {
       await Promise.all(lookups);
 
     const datasetToTerms = termsPerSource.reduce((acc, result: TermsResult) => {
-      const dataset = this.catalog.getDatasetByDistributionIri(
+      let dataset = this.catalog.getDatasetByDistributionIri(
         result.distribution.iri
       )!;
-      acc.set(dataset, result);
+      if (result instanceof Terms) {
+        const termsResult =
+          (acc.get(dataset) as Terms) ?? new Terms(result.distribution, []);
+        for (const term of result.terms) {
+          if (term.datasetIri !== undefined) {
+            const termsDataset = this.catalog.getDatasetByIri(
+              new IRI(term.datasetIri.value)
+            );
+            if (termsDataset !== undefined) {
+              dataset = termsDataset;
+              irisToDataset.set(term.id.value, dataset);
+            }
+          }
+          termsResult.terms.push(term);
+        }
+        acc.set(dataset, termsResult);
+      } else {
+        const dataset = this.catalog.getDatasetByDistributionIri(
+          result.distribution.iri
+        )!;
+        acc.set(dataset, result);
+      }
```

Comment:
This added complexity is anything but elegant, so we may want to refactor later.